core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/AvroUpsertMonitor.scala

```diff
@@ -256,23 +262,101 @@ class AvroUpsertMonitorActor(
+      val batchStream = entityUpdateDefinitionStream.chunkN(batchSize)
+
+      // convenience method to encapsulate the call to EntityService's batchUpdateEntitiesInternal
+      def performUpsertBatch(idx: Long, upsertBatch: Seq[EntityUpdateDefinition]): Future[Traversable[Entity]] = {
+        logger.info(s"inserting batch #$idx of ${upsertBatch.size} entities for jobId ${jobId.toString} ...")
+        entityService.apply(userInfo).batchUpdateEntitiesInternal(workspaceName, upsertBatch, upsert = true)
+      }
+
+      // upsert each chunk
+      val upsertFuturesStream = batchStream.zipWithIndex.map {
+        case (chunk, idx) => toFutureTry(performUpsertBatch(idx, chunk.toList))
+      }
+
+      // wait for all upserts
+      Future.sequence(upsertFuturesStream.compile.toList.unsafeRunSync()) map { upsertResults =>
+
+        val numSuccesses:Int = upsertResults.collect {
+          case Success(entityList) => entityList.size
+        }.sum
+
+        val failureReports:List[RawlsErrorReport] = upsertResults collect {
+          case Failure(regrets:RawlsExceptionWithErrorReport) => regrets.errorReport.causes
+        } flatten
+
+        // fail if nothing at all succeeded
+        if (numSuccesses == 0) {
+          // this could be a LOT of error reports, we don't want to send an enormous packet back to the caller.
+          // Cap the failure reports at 100.
+          val failureReportsForCaller = failureReports.take(100)
+          val additionalErrorString = if (failureReports.size > failureReportsForCaller.size) {
+            "; only the first 100 errors are shown."
+          } else {
+            "."
+          }
+          throw new RawlsExceptionWithErrorReport(RawlsErrorReport(StatusCodes.BadRequest,
+            s"All entities failed to update. There were ${failureReports.size} errors in total$additionalErrorString",
+            failureReportsForCaller))
+        }
+
+        val elapsed = System.currentTimeMillis() - startTime
+        logger.info(s"upsert process for $jobId succeeded after $elapsed ms: $numSuccesses upserted, ${failureReports.size} failed.")
+
+        ImportUpsertResults(numSuccesses, failureReports)
+
+      } recoverWith {
+        // recoverWith for any uncaught errors in Futures. Potential for retry here, in case of transient failures
+        case ex:Exception =>
+          val elapsed = System.currentTimeMillis() - startTime
+          logger.error(s"upsert process for $jobId FAILED after $elapsed ms: ${ex.getMessage}")
+          acknowledgeMessage(ackId) // just in case
+          Future.failed(ex)
+      }
+    } catch {
+      // try/catch for any synchronous exceptions, not covered by a Future.recover(). Potential for retry here, in case of transient failures
+      case ex:Exception =>
+        val elapsed = System.currentTimeMillis() - startTime
+        logger.error(s"upsert process for $jobId FAILED after $elapsed ms: ${ex.getMessage}")
+        acknowledgeMessage(ackId) // just in case
```

Comment:
We're calling acknowledgeMessage 3 times - once in the Try, once in the recoverWith and once in the catch. Could we call acknowledgeMessage once before the Try instead? Does it need to happen after getting the upsertStream?

Reply:
The first call, just after getting the stream, is where we originally *wanted* to call acknowledgeMessage. The calls in recover and catch are there just to be super safe ... if something goes wrong unexpectedly that we didn't otherwise plan for, we should ack the message so we don't have all these repeated "double delivery" log messages. That said ... lmk if you think we should _remove_ the calls in catch/recover, so that we would retry on those types of failures. I'm going to merge the PR as-is to move on to manual testing in dev, but could easily do a follow-on PR that changes where we ack.